### PR TITLE
Feature/unit links happy path

### PR DESCRIPTION
**Pull request by ameenhaq**

### CODE REVIEW

File: app/controllers/v1/SearchController.scala [hunk 1]
```diff
@@ -82,7 +82,7 @@ class SearchController @Inject() (db: DataAccess, playConfig: Configuration, lan
     httpMethod = "GET"
   )
   @ApiResponses(Array(
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply app/controllers/v1/SearchController.scala]: Yes there is two UnitLinks case class. UnitLinksUnit is the deprecated one that the old code uses - SearchController, the new one is as implied is now UnitLinks

File: app/controllers/v1/UnitLinksController.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import uk.gov.ons.sbr.models.Period
+import uk.gov.ons.sbr.models.unitlinks.{ UnitId, UnitLinks, UnitType }
+
+import controllers.v1.ControllerResultProcessor._
+import repository.UnitLinksRepository
+import play.api.libs.concurrent.Execution.Implicits.defaultContext
+
+class UnitLinksController @Inject() (repository: UnitLinksRepository) extends Controller {
+
+  def retrieveUnitLinksWithPeriod(id: String, periodOptStr: String, unitTypeStr: String): Action[AnyContent] = Action.async {
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply app/controllers/v1/UnitLinksController.scala]: Updated in UnitLinks routing validation branch

File: app/repository/hbase/unitlinks/UnitLinksRowKey.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  val unitLinksRowKeyLength = 3
+
+  val unitIdIndex = 0
+  val unitTypeIndex = 1
+  val unitPeriodIndex = 2
+
+  val split: (String) => List[String] = (rowKey: String) =>
+    rowKey.split(RowKeyDelimiter).toList
+
+  def apply(id: UnitId): String =
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply app/repository/hbase/unitlinks/UnitLinksRowKey.scala]: Removed completely. Not needed for this story endpoint, but will consider the point for future endpoint that uses it.

File: app/repository/hbase/unitlinks/UnitLinksRowMapper.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    } yield UnitLinks(UnitId(id), period, parents, children, unitType)
+
+  private def partitionMap(rawMap: Map[String, String]): (Map[String, String], Map[String, String]) =
+    rawMap.partition {
+      case (k, _) =>
+        k.startsWith(UnitParentPrefix)
+    }
+
+  private def returnNoneIfAllNotPrefixedAsChild(rawMap: Map[String, String]): Boolean = !rawMap.map {
+    case (key, _) =>
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply app/repository/hbase/unitlinks/UnitLinksRowMapper.scala]: Used for forAll such that Map m should satisfy p: (k, _) => k.startsWith(prefix) = Boolean

File: app/repository/hbase/unitlinks/UnitLinksRowMapper.scala [hunk 5]
```diff
@@ -111,17 +94,21 @@ object UnitLinksRowMapper extends RowMapper[UnitLinks] with LazyLogging {
...
-      logger.warn(s"Failure to produce UnitLinks, caused by rowKey [${partitionedKey.mkString}] invalid length " +
-        s"[${partitionedKey.length}] when expected [$unitLinksRowKeyLength]")
-      false
-    } else true
+    }
+    !ifChildrenOrParentsIsEmpty
+  }
+
+  private def returnNoneWhenRowKeyIsNotOfLength(partitionedKey: List[String]): Boolean = {
+    val checkPartitionedRowKeySize = partitionedKey.length == numberOfUnitLinksRowKeyComponents
```
Comment (nigelhp) [inline app/repository/hbase/unitlinks/UnitLinksRowMapper.scala]: rowKeyHasRequiredComponents?

File: test/repository/hbase/unitlinks/UnitLinksRowMapperSpec.scala [hunk 6]
```diff
@@ -25,7 +25,7 @@ class UnitLinksRowMapperSpec extends FreeSpec with Matchers with MockFactory {
     )
     val ChildrenMapStr = Map(
       UnitChildPrefix + SampleCompaniesHouseChildId -> CompaniesHouse,
-      UnitChildPrefix + SamplePayAsYouEarnChildId -> PayeAsYourEarnTax
+      UnitChildPrefix + SamplePayAsYouEarnChildId -> PayAsYourEarnTax
```
Comment (nigelhp) [inline test/repository/hbase/unitlinks/UnitLinksRowMapperSpec.scala]: YouEarn rather than YourEarn?

**Pull request merged**